feat: add device service migration guide (#503)
edgexfoundry/edgex-docs: docs_src/microservices/device/V2Migration.md

    )
    ```

#### CommandValue
`CommandValue` is redesigned to be more simple and straightforward. A single `Value` with `interface{}` type is able to accommodate reading value of supported type.
As a result, you might notice the original API to create CommandValue is no longer working.
In V2 we refactor all those API functions to create CommandValue of different type to a generic function:

!!! example "Create CommandValue with string Type"
    ```go
    cv, err := models.NewCommandValue(deviceResourceName, v2.ValueTypeString, "foobar")
    if err != nil {
        ...
    }
    cv.Origin = time.Now().Unixnano()
    cv.Tags["foo"] = "bar"
    ```

The 3rd argument in the function must be able to cast into the Type defined in 2nd argument otherwise there will be error.
See [Data formats](../../design/adr/device-service/0011-DeviceService-Rest-API.md#data-formats) for supported data type in EdgeX.

Device Service also supports [Event Tagging](../general/index.md), the tags on the CommandValue will be copied to Event.

### Device Profiles
See [Device Profile Reference](profile/Ch-DeviceProfileRef.md) for details, SDK now allows both YAML and JSON format.

#### Device Resource

> `properties` field is simplified in device resource.

Reviewer: Add some statements:

- `units` becomes a single string field and it's optional.
- Float32 and Float64 are both only represented in eNotation.  Base64 encoding is removed, so there is no `floatEncoding` field anymore.

# V2 Migration Guide

!!! edgey "EdgeX 2.0"
For the EdgeX 2.0 (Ireland) release there are many backward breaking changes. These changes require custom Device Services and custom device profiles to be migrated. This section outlines the necessary steps for this migration.

## Custom Device Services

### Configuration

The migration of any Device Service's configuration starts with migrating configuration common to all EdgeX services. See the [V2 Migration of Common Configuration](../../configuration/V2MigrationCommonConfig) section for details. The remainder of this section focuses on configuration specific to Device Services.

#### Device
1. Remove `ImitCmd`, `ImitCmdArgs`, `RemoveCmd` and `RemoveCmdArgs`
2. Add `UseMessageBus` to determine events should be published to MessageBus or sent by REST call.

> 3. Add `DevicesDir` and `ProfilesDir` as an indication of where to load the device profiles and pre-defined devices. Convention is to put them under `/res` folder:

Reviewer 2: Please provide example TOML with these new settings.

```
(where binary resides ...)
+- res
|  +- devices
|    +- device1.toml
|    +- device2.toml
|  +- profiles
|    +- profile1.yml
|    +- profile2.yml
|  +- configuration.toml
|  +- ...
+- main.go

```

#### MessageQueue

> See the [Device Service MessageQueue](../../design/adr/013-Device-Service-Events-Message-Bus.md#device-services) section for details

Reviewer 2: Add that this is new for EdgeX 2.0. I am ok with referencing the ADR for more details, but not for the complete migration. Please add description and example TOML.

### Code

#### Dependencies

> You first need to update all the dependencies to v2 major release: update `go.mod`, all the import path and data models.

Reviewer 2: Nice, I missed this one in App Service. We add it.

#### CommandValue
You might then notice the original API to create CommandValue is no longer working.
In V2 we refactor all those API functions to create CommandValue of different type to a simple generic function:

!!! example "Create CommandValue with string Type"
    ```go
    cv, err := models.NewCommandValue(deviceResourceName, v2.ValueTypeString, "foobar")
    if err != nil {
        ...
    }
    cv.Origin = time.Now().Unixnano()
    ```

The 3rd argument in the function must be able to cast into the Type defined in 2nd argument otherwise there will be error.
See [Data formats](../../design/adr/device-service/0011-DeviceService-Rest-API.md#data-formats) for supported data type in EdgeX.

### Device Profiles

> See [V2 Device Profile Reference](profile/V2DeviceProfileRef.md) for details, SDK now allows both YAML and JSON format.

Reviewer 2: The referenced document doesn't describe the differences to assist in migration. Please add description here how to migrate profiles.

Reviewer 2: Still need to detail what someone needs to change from V1 to V2 to migrate a profile.

### Pre-defined Devices

> In V2 pre-defined devices should have their own file, SDK allows both TOML and JSON format.

Reviewer 2: ```suggestion
In V2 pre-defined devices are in their own file, SDK allows both TOML and JSON format.
```

!!! example "Pre-defined devices"
    ```toml
    [[DeviceList]]
    Name = 'Simple-Device01'
    ProfileName = 'Simple-Device'
    Description = 'Example of Simple Device'
    Labels = [ 'industrial' ]
      [DeviceList.Protocols]
        [DeviceList.Protocols.other]
        Address = 'simple01'
        Port = '300'
      [[DeviceList.AutoEvents]]
      Interval = '10s'
      OnChange = false
      SourceName = 'Switch'
      [[DeviceList.AutoEvents]]
      Interval = '30s'
      OnChange = false
      SourceName = 'Image'
    ```

> Notice that we renamed some fields:  

Reviewer 2: Nice!
